File: py/portfolio.py
from money import Money
from bank import Bank


class Portfolio:
    def __init__(self):
        self.__moneys = []

    def add(self, *moneys: tuple[Money]):
        self.__moneys.extend(moneys)

    def evaluate(self, bank: Bank, currency: str):
        total = 0.0
        failures = []
        for m in self.__moneys:
            try:
                total += bank.convert(m, currency).amount
            except Exception as e:
                failures.append(e)

        if not failures:
            return Money(total, currency)

        failure_message = ",".join(i.args[0] for i in failures)
        raise Exception("Missing exchange rate(s):[" + failure_message + "]")

File: js/portfolio.js
const Money = require("./money");
const exchangeRates = require("./rates");

class Portfolio {
  constructor() {
    this.moneys = [];
  }

  add(...moneys) {
    this.moneys = this.moneys.concat(moneys);
  }

  evaluate(bank, currency) {
    let failures = [];
    let total = this.moneys.reduce((sum, money) => {
      try {
        let converted = bank.convert(money, currency);
        return sum + converted.amount;
      } catch (e) {
        failures.push(e.message);
        return sum;
      }
    }, 0);
    if (!failures.length) {
      return new Money(total, currency);
    }
    throw new Error("Missing exchange rate(s):[" + failures.join() + "]");
  }
}

module.exports = Portfolio;
